Generic weight averaging callback that supports EMA

senarvi opened the pull request:
A callback that updates an [AveragedModel](https://pytorch.org/docs/stable/generated/torch.optim.swa_utils.AveragedModel.html) after every training step

## What does this PR do?

This is similar to the existing [StochasticWeightAveraging](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.StochasticWeightAveraging.html) callback, but wraps the [AveragedModel](https://pytorch.org/docs/stable/generated/torch.optim.swa_utils.AveragedModel.html) class from PyTorch. Reduced code duplication means easier maintenance. Also, any averaging function can be used. By default, the callback does averaging on every step, but this can be customized by overriding the ``should_update(step_idx, epoch_idx)`` method.

Fixes #10914

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs) => Discussed in issue #10914
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request? => There are none.
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

<!--

Did you have fun?

Make sure you had fun coding 🙃

-->


<!-- readthedocs-preview pytorch-lightning start -->
----
📚 Documentation preview 📚: https://pytorch-lightning--20545.org.readthedocs.build/en/20545/

<!-- readthedocs-preview pytorch-lightning end -->

lantiga commented:
Hey @senarvi, this looks great!

I saw you already added support for saving and resuming which is great. There are many scenarios there (save every n steps, time-based, every epoch, etc) let's make sure we cover them all (for inspiration, we added quite a few tests here https://github.com/Lightning-AI/pytorch-lightning/pull/20379)

> we could still have different callbacks ("StepwiseAveragingCallback" and "EpochwiseAveragingCallback")

No I think it's better to have one with configurable averaging flags, more lightning-esque

> Constructs the AveragedModel with use_buffers=True, so that an extra step is not needed for updating the batch normalization statistics. StochasticWeightAveraging performs an extra step in the end. Consequently the implementation is significantly more complex and it's difficult to make sure that it works in all cases. Should we add this as an option in this class too?

I think this is ok, but my doubt with forcing `use_buffers` to be true is what happens when a user has a module with buffers in it that are not meant to be averaged. I guess at that point they will probably be the same over time (e.g. the RoPE cache), but that's not really a guarantee.

Wdyt about this? I don't necessarily want to make the implementation more complex, so this is just for discussion.

> Updates the average model after every step. StochasticWeightAveraging updates the average model after every epoch, and I recall that the original paper updated it only at certain points (the learning rate minima). I guess it would be nice to be able to select whether the average model will be updated after every step, after every epoch, or after certain epochs. Then we would need only one callback and we could remove the StochasticWeightAveraging callback, but would it make this class too complex?

It would be nice to make it configurable, and probably users will want to get to some minimum and then start averaging. The criteria to do so may be very bespoke, so maybe allowing the user to implement a custom hook to decide whether to start averaging or whether to average at a given step would be super handy. Otherwise I'm expecting users will train for some time, save a checkpoint, then reload with this callback added to the trainer and start averaging. Which is totally fine but it requires you to stop and resume.

Regarding removing the StochasticWeightAveraging callback, I don't necessarily see that happening. We have a pretty strong commitment to backward compatibility at this point, so keeping that in with a notice to just use this one will not hurt.

cyanic-selkie commented:
Hi, I have a couple questions.

1. You added the `on_validation_epoch_start` and `on_validation_epoch_end` hooks to swap the weights, but shouldn't the same happen for `test`?
2. In my current workflow I have a separate script that does the model exporting to ONNX. It's short, and really the only `Lightning` specific thing is the `MyLightningModule.load_from_checkpoint(...)` method. Since the averaged weights are a part of the callback, I would have to instantiate the trainer for the weights to be loaded. And even then, I wouldn't have a function I could call to explicitly swap the weights (since `_swap_weights` is private and not really accessible). So, my question is, can we have some sort of an API, outside of the trainer, that can load the averaged weights instead of the regular weights? Perhaps adding some sort of a parameter to the `load_from_checkpoint` method?

cyanic-selkie commented:
@senarvi Ah! Thanks for the clarification, I should've checked the code out more carefully. I tried your branch out on a quantization aware training enabled model with ONNX export at the end and everything is working beautifully! I hope this gets merged quickly.

senarvi commented:
The user can now provide either the `update_on_step` or the `update_on_epoch` argument. (In theory also both.) It should be a function that takes the step/epoch number and returns `True` if the average model should be updated at that point of time.

For example:
```python
update_on_step = lambda x: x > 100
```
or
```python
update_on_epoch = lambda x: x in (3, 5, 7)
```

Using `update_on_epoch`, SWA should be possible. I added one unit test for SWA.

I tested EMA in an actual learning task and it gave an improvement, so I'm starting to be more confident that this works.

I think the biggest question that is still left is whether it's a problem that we force `use_buffers=True`. It would be nice if we could provide the option to instead call [update_bn()](https://pytorch.org/docs/stable/optim.html#torch.optim.swa_utils.update_bn) after training and we wouldn't have to duplicate any of that code. That function takes a data loader and iterates through the data. I can imagine that passing the Trainer's data loader might not work in all cases. We could also leave calling this function to the user.

`StochasticWeightAveraging` [increments the number of epochs](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/callbacks/stochastic_weight_avg.py#L174) in `on_fit_start()` and during the extra epoch [disables the backward pass](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/callbacks/stochastic_weight_avg.py#L258). I could also copy the code from that class, but there are some details that I don't understand, and I'm not that excited of copying code that I don't fully understand.

@tchaton I think you contributed the `StochasticWeightAveraging` callback, maybe you have some insight?

cyanic-selkie commented:
Is there anything blocking this from being merged?

senarvi commented:
I marked this pull request ready for review. There were no comments whether it's a problem that we force `use_buffers=True`. Would it make sense to merge this now and perhaps introduce such option later based on the feedback that we receive?

lantiga commented on src/lightning/pytorch/callbacks/weight_averaging.py:
```diff
@@ -0,0 +1,288 @@
+        avg_fn: The averaging function used to update the parameters. The function must take in an
+            :class:`AveragedModel` parameter, a current model parameter, and the number of models already averaged. If
+            ``None``, an equally weighted average will be used.
+        update_on_step: A function that takes the number of optimizer steps taken, and returns ``True`` if the average
```
An alternative (or an addition) could be two methods (`self.update_on_step` and `self.update_on_epoch`)  that the user needs to override. This would allow a user to get a hold of the module and the trainer, and condition saving on richer information like the value of the loss.

senarvi replied:
@lantiga That's true. Overriding some method seems to be a common way of customizing callbacks in Lightning.

I changed it like this. However, there's just one method for the user to override, `should_update(self, step_idx: Optional[int] = None, epoch_idx: Optional[int] = None)`. (Otherwise, if the user would override `update_on_epoch()` and not override `update_on_step()`, the callback would still update on every step.)

lantiga commented on src/lightning/pytorch/callbacks/weight_averaging.py:
```diff
@@ -0,0 +1,288 @@
+        checkpoint["state_dict"] = {
+            name[7:]: value for name, value in average_model_state.items() if name.startswith("module.")
+        }
+        checkpoint["averaging_state"] = {
```
```suggestion
        checkpoint["averaged_state"] = {
```
I get that it might be still "averaging" : ), but it's in fact "averaged" up to the current iterations. We can called it "average" model if it sounds better.

senarvi replied:
@lantiga the name is a bit confusing, but it means the state of the averaging process, not the average model. This includes the state variables of the `AveragedModel` class, excluding the `module` (i.e. `n_averaged`). The average model is saved in `state_dict`, so whatever we'll do with the checkpoint, we'll use the average model. The current model state is saved in `current_model_state`, so that we can continue training with the WeightAveraging callback from the previous state. If you have a less confusing name for the "averaging state variables, excluding the averaged model parameters", I can change it.

lantiga commented on src/lightning/pytorch/callbacks/weight_averaging.py:
```diff
@@ -0,0 +1,288 @@
+
+        """
+        if self._average_model is None:
+            raise Exception("Trying to load a checkpoint, but no average model (outside fit). Don't know what to do.")
```
This is hard to understand for a user if they don't know the details of the callback.

```suggestion
            raise Exception("Trying to load a checkpoint using the WeightAveraging callback outside the `fit` stage. The WeightAveraging callback can only be used in the `fit` stage.")
```

I'm wondering: instead of raising we could just load the average model e.g. for predict. This will avoid forcing users to remove the callback from the Trainer.

senarvi replied:
@lantiga I guess I just wasn't sure in which situation this callback would be called outside fit, but yes, if the user calls `Trainer.validate/test/predict(ckpt_path=...)`, I believe this will be called and the best thing to do would be to load the average model. The average model will be loaded if we don't do anything. Maybe just display a warning in that case.

I don't know if `on_save_checkpoint()` is ever called outside fit. What do you think is the best action there?

senarvi replied:
@lantiga This is what I did. Please check if you think the messages are clear now.

lantiga commented on tests/tests_pytorch/callbacks/test_weight_averaging.py:
```diff
@@ -0,0 +1,288 @@
+    assert trainer.lightning_module == model
+
+
+def _train_and_resume(tmp_path: str, crash_on_epoch: int, use_ddp: bool = False) -> None:
```
This tests that we can crash and resume, but afaict it doesn't test whether the resulting averaging is equivalent. We can harden this in a subsequent PR, but it is important to know for sure that averaging works if I stop training and resume while averages are being taken, irrespective of where I stop and resume in the lifecycle.

senarvi replied:
I'll try to still improve the test.

senarvi replied:
@lantiga now I test that after stopping and resuming we get the same final model. The parameters are not identical - I have to use `atol=0.001` - but they are close enough so that I think that the difference comes from some random change instead of a bug in restoring the checkpoint. I don't know what could cause the difference, though. I pass `deterministic=True` to Trainer. I'm curious if you have some ideas, or if you think that that's close enough.

lantiga reviewed (commented):
Solid contribution @senarvi! I added a few comments (most are quick to address, let me know what you can do here vs follow up PR), but overall looks great.

lantiga commented:
BTW: I think it's totally fine to merge this as is and open an issue to gather discussions about averaging buffers.

The other question I have (for the future) is related to fitting both models on GPU. It may make sense to give the ability to keep the AveragedModel on a different device (e.g. `cpu`) to keep the callback usable with larger models.

senarvi commented:
> The other question I have (for the future) is related to fitting both models on GPU. It may make sense to give the ability to keep the AveragedModel on a different device (e.g. `cpu`) to keep the callback usable with larger models.

There's a `device` argument already, and actually the default is `cpu` - as with `StochasticWeightAveraging`.

senarvi commented:
> Hi! Thanks for this great PR. The current implementation only leverages `avg_fn` argument should it also consider the in-place version `multi_avg_fn` ?

I think we could just pass `**averaged_model_kwargs`. I'll look into it over the weekend.

senarvi commented:
Now any extra keyword arguments will be passed to the `AveragedModel` constructor, so you can provide either `avg_fn` or `multi_avg_fn`.

You can also set `use_buffers=False`, if you don't want the model buffers to be averaged. In that case it's your responsibility to call `torch.optim.swa_utils.update_bn()` afterwards, if needed.

By the way, @h2o64 would you mind testing that the callback works in your use case and leaving a review? Testing it is pretty simple, because the entire implementation is contained in one class. The easiest way to test it is to simply copy [the WeightAveraging class from the source repo](https://github.com/senarvi/lightning/blob/generic-weight-averaging/src/lightning/pytorch/callbacks/weight_averaging.py#L33) to your project.

cyanic-selkie commented:
A respectful bump :)

senarvi commented:
Hi @cyanic-selkie ! Thanks for bringing this into attention. I fixed the merge conflict in CHANGELOG. I also updated the documentation (training tricks). So, from my part this is ready and I've been using it succesfully. You also said that it's working for you. Some tests fail, but I think the problem is not in this PR. @lantiga does this look like it could be merged?

catalpaaa commented:
Works well for me too <3

A small suggestion, can we find a better way to print `Loading the average model parameters for validation.` and `Recovering the current model parameters after validation.`. Those are making `tqdm` cry and piss.

![image](https://github.com/user-attachments/assets/2b17e988-2003-44ae-85b9-adf31ce52779)

Borda commented:
will check it shortly :)

senarvi commented:
Thank you @catalpaaa for checking the PR! I removed the logging. I think it's not needed anymore. Originally, my biggest fear was that the weights are not transferred correctly, which could go unnoticed. And actually... I was kind of right.

The `test_ema_resume` test was failing in the CI pipeline. I had to set the tolerance to a suspiciously high number when comparing the model weights. So I started once again looking into where that difference comes from.

`test_ema_resume` trains two models, one without interruption and one with an intentional crash and recovery from a checkpoint after *N* epochs. I noticed that the weights that I load from the checkpoint are not the same that the model has in the beginning of epoch *N+1*. It appears that Lightning has already loaded the model weights from "state_dict" when entering the `on_load_checkpoint()` callback. I had assumed that I can swap the "current_model_state" to "state_dict" in the callback, but in fact I have to reload the model state from "current_model_state".

I'm super happy that I found this bug. Now, I think, the only test that's failing is `tests_pytorch/callbacks/test_pruning.py::test_pruning_callback_ddp[True-True]` in `PyTorch | oldest`. I don't find other error except `Bash exited with code '1'`.

catalpaaa commented:
When training with TPUs, I add 

```python
callbacks.append(WeightAveraging(avg_fn=get_ema_avg_fn(0.9999)))
```

to my call backs. I will get the following errors:

```
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/queues.py", line 244, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
AttributeError: Can't pickle local object 'get_ema_avg_fn.<locals>.ema_update'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/root/Theia/train.py", line 139, in <module>
    main()
  File "/root/Theia/train.py", line 135, in main
    trainer.fit(model, datamodule=data)
  File "/usr/local/lib/python3.10/dist-packages/lightning/pytorch/trainer/trainer.py", line 567, in fit
    call._call_and_handle_interrupt(
  File "/usr/local/lib/python3.10/dist-packages/lightning/pytorch/trainer/call.py", line 47, in _call_and_handle_interrupt
    return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/lightning/pytorch/strategies/launchers/xla.py", line 98, in launch
    process_context = xmp.spawn(
  File "/usr/local/lib/python3.10/dist-packages/torch_xla/distributed/xla_multiprocessing.py", line 39, in spawn
    return pjrt.spawn(fn, nprocs, start_method, args)
  File "/usr/local/lib/python3.10/dist-packages/torch_xla/_internal/pjrt.py", line 213, in spawn
    run_multiprocess(spawn_fn, start_method=start_method)
  File "/usr/local/lib/python3.10/dist-packages/torch_xla/_internal/pjrt.py", line 169, in run_multiprocess
    replica_results = list(
  File "/usr/local/lib/python3.10/dist-packages/torch_xla/_internal/pjrt.py", line 170, in <genexpr>
    itertools.chain.from_iterable(
  File "/usr/lib/python3.10/concurrent/futures/process.py", line 575, in _chain_from_iterable_of_lists
    for element in iterable:
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 621, in result_iterator
    yield _result_or_cancel(fs.pop())
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 319, in _result_or_cancel
    return fut.result(timeout)
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 458, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/lib/python3.10/multiprocessing/queues.py", line 244, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
AttributeError: Can't pickle local object 'get_ema_avg_fn.<locals>.ema_update'
```

The code runs fine with single GPU.

If I pull the decay function out from pytorch's `get_ema_avg_fn`:

```python
@torch.no_grad()
def ema_update(ema_param: Tensor, current_param: Tensor, num_averaged):
    decay = 0.9999
    return decay * ema_param + (1 - decay) * current_param

callbacks.append(WeightAveraging(avg_fn=ema_update))
```

Everything runs fine. Please let me know if I'm using this wrong.

senarvi commented:
Thanks @catalpaaa . You're using it correctly. I noticed the same thing with ddp_spawn, so I had to use a similar workaround in the unit tests.

The problem seems to be caused by `get_ema_avg_fn()` returning a closure. I don't know if there's anything I can do about it. If `get_ema_avg_fn` was a class instead of a function that returns a closure, like in [the unit tests](https://github.com/senarvi/lightning/blob/5deb0bbc01c7414ac50fe2635ec97898060f50ab/tests/tests_pytorch/callbacks/test_weight_averaging.py#L50), the problem would be solved. Since the whole point was to avoid duplicating code between pytorch and lightning, maybe it would be best to fix this in pytorch.

catalpaaa commented:
I guess pickle just hates function in function :(

amorehead commented:
*Another respectful bump*

amorehead commented:
Also, I've filled in the `EMAWeightAveraging` example into what may be a nice default callback configuration for users. Feel free to have a look:

```python
import torch

from torch.optim.swa_utils import get_ema_avg_fn
from typing import Optional, Union


class EMAWeightAveraging(WeightAveraging):
  """Exponential Moving Average (EMA) Weight Averaging callback."""

  def __init__(
      self,
      device: Optional[Union[torch.device, str, int]] = "cpu",
      use_buffers: bool = True,
      decay: float = 0.999,
      update_every_n_steps: int = 1,
      update_starting_at_step: int | None = 100,
      update_starting_at_epoch: int | None = 0,
      **kwargs: Any,
  ):
      super().__init__(
          device=device,
          use_buffers=use_buffers,
          **kwargs,
          avg_fn=get_ema_avg_fn(decay=decay),
      )

      self.update_every_n_steps = update_every_n_steps
      self.update_starting_at_step = update_starting_at_step
      self.update_starting_at_epoch = update_starting_at_epoch

  def should_update(self, step_idx: int | None = None, epoch_idx: int | None = None):
      """Decide when to update the model weights. By default, this is done
      every step after the first 100 steps.

      Args:
          step_idx: The current step index.
          epoch_idx: The current epoch index.
      Returns:
          bool: True if the model weights should be updated, False otherwise.
      """
      every_n_steps_update = (
          step_idx is not None
          and self.update_every_n_steps is not None
          and step_idx % self.update_every_n_steps == 0
      )
      step_update = (
          (step_idx is not None)
          and (self.update_starting_at_step is not None)
          and (step_idx >= self.update_starting_at_step)
      )
      epoch_update = (
          (epoch_idx is not None)
          and (self.update_starting_at_epoch is not None)
          and (epoch_idx >= self.update_starting_at_epoch)
      )

      return every_n_steps_update or step_update or epoch_update
```

senarvi commented:
That's a very good catch @amorehead ! Thanks for testing this before it's merged.

I wasn't familiar with the `configure_model` hook, so I wanted to understand that we're using it correctly. According to the [documentation](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#configure-model), it's called in a "strategy and precision aware context", so that when using a sharded strategy, the model is sharded instantly. Normally it's called in [`_call_configure_model()`](https://github.com/Lightning-AI/pytorch-lightning/blob/1af7f73d4b1043b8be60babdfbd15637c96a0374/src/lightning/pytorch/trainer/call.py#L114-L128) like this:

```python
with (
    trainer.strategy.tensor_init_context(),
    trainer.strategy.model_sharded_context(),
    trainer.precision_plugin.module_init_context(),
):
    _call_lightning_module_hook(trainer, "configure_model")
```

Should we call `_call_configure_model(trainer)` instead of `pl_module.configure_model()`? Otherwise, I guess the model is not sharded at this point.

amorehead commented:
@senarvi, you are right! I just noticed that the model will not be sharded unless Lightning's private method `lightning.pytorch.trainer.call._call_configure_model` is called instead of `pl_module.configure_model`. However, I don't think this will be enough to fully support model-parallel training strategies (such as FSDP2, which I'm currently trying to test with this callback), since each GPU rank will contain only a portion of the model weights (and there's currently only a single `_average_model` instance being updated).

senarvi commented:
@amorehead , if I understand correctly, each GPU rank is running in its own process that contains a portion (shard) of the model weights and an `AveragedModel` instance. Because of issue 1, `AveragedModel` is not able to copy the parameters of an FSDP2 model, but if that issue is solved, the `AveragedModel` of each GPU rank would contain averaged values of the corresponding shard. In theory, we should be able to construct the final averaged model by gathering the `AveragedModel` weights from all GPU ranks, right? This should be done before saving a checkpoint, or we'll end up saving only the shard of rank 0.

The other option is to call `configure_model()` without `model_sharded_context`, then construct the `AveragedModel`, and finally shard both models. Your suggesting that if we simply store the `AveragedModel` as a property of `pl_module`, it gets sharded automatically. Two questions come to my mind:

1) Is the sharding of the averaged parameters guaranteed to be identical to the sharding of the original parameters?
2) If you need to use FSDP, you probably want the averaged parameters to be on CPU, right? I think this would cause the averaged parameters to be stored on the same device as the original parameters.

amorehead commented:
@senarvi, for the time being, it seems like the [reference implementation](https://github.com/pytorch/pytorch/blob/7b94d934313320eff91236540d37021d62bc8984/test/distributed/fsdp/test_fsdp_mixed_precision.py#L1353) mentioned in [this GitHub issue](https://github.com/pytorch/pytorch/issues/117742) is a standardized (PyTorch-tested) way of running EMA weight updates with FSDP. At first glance, with this reference implementation, I'm not sure if the EMA weights can be stored on any device other than a GPU (for sake of sharding). Also, this GitHub issue makes me realize that no matter what, one will have to instantiate the full set of model weights, either to initialize the (full) EMA weights or to call `FSDP.summon_full_params` for EMA weight updates after every optimizer step.

yhao-z commented:
When i use this WeightAveraging Callback along with ModelCheckpoint Callback, i noticed that `save_last=True` in ModelCheckpoint Callback fails. I don't realy know why, but it just happened. plz check, thx a lot.

amorehead commented:
@senarvi, I think it would be best practice to add or use a Lightning hook that accomplishes this weight gathering in a standardized way, then we can just reference that hook in this callback once its implemented. For the time being, just like @cyanic-selkie said, I think this PR is more than good enough as is to merge, since it works well for DDP training.

amorehead commented:
@yhao-z, make sure you are not using the `configure_model` hook with `DDPStrategy`. Otherwise, the `WeightAveraging` callback may fail to correctly wrap your original model weights and thus fail to save a last checkpoint.

senarvi commented:
I agree, better to do this in small steps and it's probably better that the support for sharded strategies is written by someone who actually uses them. But something like what @amorehead said, a standardized way to gather the weights sounds like best practice.

For now, I just wanted to make sure that if someone tries to use weight averaging with `configure_model()`, it won't fail silently and produce incorrect results. This is what I did:

- `setup()` will [call configure_model()](https://github.com/senarvi/lightning/blob/generic-weight-averaging/src/lightning/pytorch/callbacks/weight_averaging.py#L155) if it's overridden, but issues a warning. Likely runs out of memory too.
- I added a [unit test](https://github.com/senarvi/lightning/blob/generic-weight-averaging/tests/tests_pytorch/callbacks/test_weight_averaging.py#L276) that checks that the `AveragedModel` is constructed correctly if `configure_model()` is overridden, so it should work if it doesn't run out of memory. (The CI pipeline runs out of memory if I use two GPUs, so I couldn't include that.)
- I added a [note](https://github.com/senarvi/lightning/blob/generic-weight-averaging/src/lightning/pytorch/callbacks/weight_averaging.py#L59) in the documentation that weight averaging doesn't work with sharded models.

@amorehead How does that sound to you?

@yhao-z was your issue also caused by overriding `configure_model()`? If not, please send me a minimal example that I can use to reproduce the error. For example, here's a minimal example that uses `ModelCheckpoint(save_last=True)`, but doesn't fail.

```python
import torch
from torch import nn
from torch.optim.swa_utils import get_ema_avg_fn
from torch.utils.data import DataLoader

from lightning.pytorch import Trainer
from lightning.pytorch.callbacks import ModelCheckpoint, WeightAveraging
from lightning.pytorch.demos.boring_classes import BoringModel, RandomDataset

class EMATestModel(BoringModel):
    def __init__(self):
        super().__init__()
        layers = [nn.Linear(32, 32)]
        layers += [nn.ReLU(), nn.Linear(32, 2)]
        self.layer = nn.Sequential(*layers)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

class EMAWeightAveraging(WeightAveraging):
    def __init__(self):
        super().__init__(avg_fn=get_ema_avg_fn())

    def should_update(self, step_idx=None, epoch_idx=None):
        return (step_idx is not None) and (step_idx >= 10)


dataset = RandomDataset(32, 64)
dataloader = DataLoader(dataset, batch_size=2)
model = EMATestModel()
trainer = Trainer(callbacks=[EMAWeightAveraging(), ModelCheckpoint(dirpath=".", save_last=True)], max_epochs=10)
trainer.fit(model, dataloader)
```

senarvi commented:
@Borda if there's nothing else, all the checks passed and this could be a good time to merge.

amorehead commented:
@senarvi, thanks for updating the PR accordingly! Yes, these changes look good to me and seem reasonable for this first version of the callback.

alberthli commented:
Any timeline on when this will be merged?

senarvi commented:
All comments have been resolved. It says "Waiting on code owner review from @Borda , @ethanwharris , @justusschock , @lantiga , and/or @tchaton ." Sorry for bugging you. :) I guess we can merge if we get an approving review from one of you.

amorehead commented:
Seconded, this PR would be very useful to merge soon!

catalpaaa commented:
im 

> Hi @cyanic-selkie
> 
> During training (stage=fit), the actual LightningModule is what we update using the optimizer (I call it the current model) and an AveragedModel is maintained in the background (I call it the average model).
> 
> I assume that validation is only called during training. Before and after validation we swap the current model and the average model, so the average model will be validated.
> 
> When saving a checkpoint, we save the average model parameters in the state_dict. So if you later load the checkpoint without WeightAveraging callback and run a test or export to ONNX, you will be using the average parameters.
> 
> When training ends, we copy the average model parameters to the current model. So if you run a test or export to ONNX after training, you will be using the average parameters.
> 
> That's the idea at least. I'm not confident that I have thought about every possible corner case. It would be great if you could test that it works in your case.

I run training without any validation run, i.e. `trainer = pl.Trainer(limit_val_batches=0.0)`.

After I load each checkpoint using `.load_from_checkpoint()`, I found that every `state_dict` loaded is 100% identical even thought the training loss shows convergence. All of the actual weights are saved in `current_model_state`. I think `state-dict` was never touched after init in my case.

senarvi commented:
@catalpaaa I need to know exactly how to repeat your problem and the easiest way to do that is that you send me a minimal script that demonstrates it. When you say that every `state_dict` is identical, do you mean that you save a checkpoint after every epoch and then compare them? Like this:

```python
from pathlib import Path

import torch
from torch import nn
from torch.optim.swa_utils import get_ema_avg_fn
from torch.utils.data import DataLoader

from lightning.pytorch import Trainer
from lightning.pytorch.callbacks import ModelCheckpoint, WeightAveraging
from lightning.pytorch.demos.boring_classes import BoringModel, RandomDataset

class EMATestModel(BoringModel):
    def __init__(self):
        super().__init__()
        layers = [nn.Linear(32, 32)]
        layers += [nn.ReLU(), nn.Linear(32, 2)]
        self.layer = nn.Sequential(*layers)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

class EMAWeightAveraging(WeightAveraging):
    def __init__(self):
        super().__init__(avg_fn=get_ema_avg_fn())

    def should_update(self, step_idx=None, epoch_idx=None):
        return (step_idx is not None) and (step_idx >= 10)


dataset = RandomDataset(32, 64)
dataloader = DataLoader(dataset, batch_size=2)
model = EMATestModel()
trainer = Trainer(callbacks=[EMAWeightAveraging(), ModelCheckpoint(dirpath=".", save_top_k=-1)], max_epochs=10, limit_val_batches=0.0)
trainer.fit(model, dataloader)

for checkpoint_path in Path(".").glob("*.ckpt"):
    checkpoint = torch.load(checkpoint_path)
    print(checkpoint["state_dict"]["layer.0.weight"][:5, :5])
```

What I see is that the weights are close, but not identical:

```
tensor([[-0.0649,  0.1751,  0.1634, -0.0211,  0.1116],
        [-0.1011, -0.0956,  0.0125, -0.1022, -0.1861],
        [ 0.1548, -0.0557,  0.0618,  0.0778, -0.1194],
        [-0.1239,  0.0959, -0.1402, -0.1696,  0.1550],
        [ 0.0454, -0.0599, -0.0902,  0.0385, -0.1447]])
tensor([[-0.0647,  0.1747,  0.1629, -0.0212,  0.1116],
        [-0.1021, -0.0972,  0.0118, -0.1005, -0.1844],
        [ 0.1561, -0.0569,  0.0617,  0.0777, -0.1191],
        [-0.1238,  0.0957, -0.1403, -0.1695,  0.1554],
        [ 0.0459, -0.0613, -0.0916,  0.0384, -0.1438]])
tensor([[-0.0646,  0.1746,  0.1627, -0.0212,  0.1116],
        [-0.1024, -0.0978,  0.0115, -0.0998, -0.1836],
        [ 0.1566, -0.0573,  0.0616,  0.0776, -0.1189],
        [-0.1238,  0.0956, -0.1403, -0.1695,  0.1556],
        [ 0.0462, -0.0618, -0.0923,  0.0384, -0.1433]])
...
```

Could you modify the script so that it demonstrates your issue?

catalpaaa commented:
@senarvi 

```python
from pathlib import Path

import torch
from torch import nn, Tensor
from torch.utils.data import DataLoader
from typing import Optional

from lightning.pytorch import Trainer
from lightning.pytorch.callbacks import ModelCheckpoint, WeightAveraging
from lightning.pytorch.demos.boring_classes import BoringModel, RandomDataset


class EMATestModel(BoringModel):
    def __init__(self):
        super().__init__()
        self.layer = nn.Sequential(
            nn.Linear(32, 32),
            nn.ReLU(),
            nn.Linear(32, 2),
        )

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


class EMAAveragingFunction:
    def __init__(self, decay: float) -> None:
        self.decay = decay

    @torch.no_grad()
    def __call__(
        self, ema_param: Tensor, current_param: Tensor, num_averaged: Tensor
    ) -> Tensor:
        return self.decay * ema_param + (1 - self.decay) * current_param


class EMAWeightAveraging(WeightAveraging):
    """Exponential Moving Average (EMA) Weight Averaging callback."""

    def __init__(
        self,
        decay: float,
        use_buffers: bool = True,
        update_every_n_steps: int = 1,
        update_starting_at_step: Optional[int] = None,
        update_starting_at_epoch: Optional[int] = None,
    ):
        super().__init__(
            device=None,
            use_buffers=use_buffers,
            avg_fn=EMAAveragingFunction(decay=decay),
        )

        self.update_every_n_steps = update_every_n_steps
        self.update_starting_at_step = update_starting_at_step
        self.update_starting_at_epoch = update_starting_at_epoch

    def should_update(
        self, step_idx: Optional[int] = None, epoch_idx: Optional[int] = None
    ):
        """Decide when to update the model weights.

        Args:
            step_idx: The current step index.
            epoch_idx: The current epoch index.
        Returns:
            bool: True if the model weights should be updated, False otherwise.
        """
        if step_idx is not None:
            # Check step-based conditions only if we have a valid step_idx
            meets_step_requirement = (
                self.update_starting_at_step is None
                or step_idx >= self.update_starting_at_step
            )
            meets_step_frequency = (
                self.update_every_n_steps > 0
                and step_idx % self.update_every_n_steps == 0
            )
            if meets_step_requirement and meets_step_frequency:
                return True

        if epoch_idx is not None:
            # Check epoch-based condition only if we specify one
            meets_epoch_requirement = (
                self.update_starting_at_epoch is not None
                and epoch_idx >= self.update_starting_at_epoch
            )
            if meets_epoch_requirement:
                return True

        return False


dataset = RandomDataset(32, 64)
dataloader = DataLoader(dataset, batch_size=2)
model = EMATestModel()
trainer = Trainer(
    callbacks=[EMAWeightAveraging(decay=0.999), ModelCheckpoint(dirpath=".", save_top_k=-1)],
    max_epochs=10,
    limit_val_batches=0.0,
)
trainer.fit(model, dataloader)

for checkpoint_path in Path(".").glob("*.ckpt"):
    checkpoint = torch.load(checkpoint_path)
    print("checkpoint_path")
    print("state_dict", checkpoint["state_dict"]["layer.0.weight"][:5, :5])
    print("current_model_state", checkpoint["current_model_state"]["layer.0.weight"][:5, :5])
```

Every `state_dict` is 100% equal, while the train weight is saved in `current_model_state`. Please note that I'm training on multiple TPUs, maybe that is the problem.

catalpaaa commented:
> @catalpaaa when I run that script, for me the `state_dict`s are different. Either we have different Lightning code, or then it's the fact that we're using different hardware. I would rule out the possible reasons one by one:
> 
>     1. Please try running the script on the source branch of this pull request? That is, clone the generic-weight-averaging branch here https://github.com/senarvi/lightning/tree/generic-weight-averaging
> 
>     2. If you still see identical `state_dict`s, try training with only one TPU.
> 
>     3. If you still see identical `state_dict`s, try training on CPU.
> 
> 
> I don't have TPUs so I cannot test that.

The code is updated, thanks for the reminder.

With the dummy code, running on single TPU core dose show changes, although it still have some weird behaviour. During linear warm up epochq, i set should_update to false. If I understand correctly, that will just update state_dict normally and keep self._average_model as init.

senarvi commented:
@catalpaaa if I understand correctly, when you run the script on a single TPU, the checkpoint `state_dict` (that is, the average model) changes from epoch to epoch. But when you run it on multiple TPUs, the `state_dict` doesn't change. That would indicate a problem with distributed computing on TPUs.

When `should_update()` returns `False`, it won't update the average model. That's expected.

Someone with access to TPUs should check what's going on. Otherwise we can merge this pull request and report this bug in an issue.

The average model is updated [here](https://github.com/senarvi/lightning/blob/3dafb4c0b24745a3ca8d5a877978c2cf332ae6b3/src/lightning/pytorch/callbacks/weight_averaging.py#L180):

```python
    def on_train_batch_end(
        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", outputs: STEP_OUTPUT, batch: Any, batch_idx: int
    ) -> None:
        step_idx = trainer.global_step - 1
        if (trainer.global_step > self._latest_update_step) and self.should_update(step_idx=step_idx):
            assert self._average_model is not None
            self._average_model.update_parameters(pl_module)
            self._latest_update_step = trainer.global_step
```

The average model is not updated, if
- `on_train_batch_end()` is not called
- `trainer.global_step` is not increased
- `should_update()` returns `False`

gergol commented on src/lightning/pytorch/callbacks/weight_averaging.py:
```diff
@@ -0,0 +1,361 @@
+
+    def __init__(
+        self,
+        device: Optional[Union[torch.device, str, int]] = "cpu",
```
Shouldn't this better default to `None`? Otherwise the default behaviour will slow down the training significantly.

senarvi replied:
@gergol if `device=None`, it will try to create the averaged model in the same device where the original model is. If it's a large model, the risk is that it won't fit in that device. If `device="cpu"`, it may slow down training - how significantly, depends on how often the averaged model is updated. I don't know which value is the better default. The [StochastickWeightAveraging callback](https://github.com/Lightning-AI/pytorch-lightning/blob/afa7d56eb7d6566af1bacc644435b7bde2e50487/src/lightning/pytorch/callbacks/stochastic_weight_avg.py#L47) uses `"cpu"` and I chose to use it for consistency. On the other hand, the default for [AveragedModel](https://github.com/pytorch/pytorch/blob/134179474539648ba7dee1317959529fbd0e7f89/torch/optim/swa_utils.py#L220) is `None`, so from that perspective it would be logical to use `None`. What do you think?

senarvi commented:
@catalpaaa I checked the [PyTorch AveragedModel documentation](https://docs.pytorch.org/docs/stable/generated/torch.optim.swa_utils.AveragedModel.html) and it looks like it should work as you described.

> When [update_parameters()](https://docs.pytorch.org/docs/stable/generated/torch.optim.swa_utils.AveragedModel.html#torch.optim.swa_utils.AveragedModel.update_parameters) is called for the first time (i.e. n_averaged is 0) the parameters of model are copied to the parameters of [AveragedModel](https://docs.pytorch.org/docs/stable/generated/torch.optim.swa_utils.AveragedModel.html#torch.optim.swa_utils.AveragedModel). For every subsequent call of [update_parameters()](https://docs.pytorch.org/docs/stable/generated/torch.optim.swa_utils.AveragedModel.html#torch.optim.swa_utils.AveragedModel.update_parameters) the function avg_fn is used to update the parameters.

Also, see the example in that documentation:

```python
>>> loader, optimizer, model, loss_fn = ...
>>> swa_model = torch.optim.swa_utils.AveragedModel(model)
>>> scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer,
>>>                                     T_max=300)
>>> swa_start = 160
>>> swa_scheduler = SWALR(optimizer, swa_lr=0.05)
>>> for i in range(300):
>>>      for input, target in loader:
>>>          optimizer.zero_grad()
>>>          loss_fn(model(input), target).backward()
>>>          optimizer.step()
>>>      if i > swa_start:
>>>          swa_model.update_parameters(model)
>>>          swa_scheduler.step()
>>>      else:
>>>          scheduler.step()
>>>
>>> # Update bn statistics for the swa_model at the end
>>> torch.optim.swa_utils.update_bn(loader, swa_model)
```

That's what we're doing, right? We create the `AveragedModel` in `setup()` and then we call [update_parameters()](https://docs.pytorch.org/docs/stable/generated/torch.optim.swa_utils.AveragedModel.html#torch.optim.swa_utils.AveragedModel.update_parameters) when `should_update()` returns `True`.

catalpaaa commented:
@senarvi Your example shows the same behaviour of the current implementation.

The behaviour I was referencing can be found in the [torchvision trainer](https://github.com/pytorch/vision/blob/6aee5eddc8f9b397b45d076bd336c030518b7626/references/classification/train.py#L48). The trainer `model_ema.n_averaged.fill_(0)` after every `update_parameters()` before `ema_start`, which simply [copy the model parameters to ema model](https://github.com/pytorch/pytorch/blob/134179474539648ba7dee1317959529fbd0e7f89/torch/optim/swa_utils.py#L194). This make sure the first ema step is performed on the latest model parameters, instead of the initial values.

senarvi commented:
@catalpaaa the Torchvision code first calls `update_parameters()` and then, during warmup, resets `n_averaged` to zero.

Our code only calls `update_parameters()` when `should_update()` returns `True`. When that happens for the first time, `n_averaged` is zero and `AveragedModel` will copy the weights.

So, during the warmup (before `should_update()` first returns `True`), the averaged model contains the initial weights. But they will be abandoned and not used for averaging.

Borda commented:
@SkafteNicki mind have look, pls 🦩

senarvi commented on docs/source-pytorch/advanced/training_tricks.rst:
```diff
@@ -50,23 +50,44 @@ Read more about :ref:`Configuring Gradient Clipping <configure_gradient_clipping
+    class EMAWeightAveraging(WeightAveraging):
+        def __init__(self):
+            super().__init__(avg_fn=get_ema_avg_fn())
+        def should_update(self, step_idx=None, epoch_idx=None):
```
I added a bit more verbose description.

SkafteNicki commented on docs/source-pytorch/starter/introduction.rst:
```diff
@@ -252,7 +252,7 @@ Enable advanced training features using Trainer arguments. These are state-of-th
 
    # access the latest state of the art techniques
-   trainer = L.Trainer(callbacks=[StochasticWeightAveraging(...)])
+   trainer = L.Trainer(callbacks=[WeightAveraging(...)])
```
@senarvi thanks for giving the background info. Lets just leave as is.

SkafteNicki commented on docs/source-pytorch/advanced/training_tricks.rst:
```diff
@@ -50,23 +50,44 @@ Read more about :ref:`Configuring Gradient Clipping <configure_gradient_clipping
+    from torch.optim.swa_utils import get_ema_avg_fn
+
+    # Enable Exponential Moving Average after 100 steps
+    class EMAWeightAveraging(WeightAveraging):
+        def __init__(self):
+            super().__init__(avg_fn=get_ema_avg_fn())
+        def should_update(self, step_idx=None, epoch_idx=None):
+            return (step_idx is not None) and (step_idx >= 100)
+    trainer = Trainer(callbacks=EMAWeightAveraging())
```
Yes, perfectly clear

SkafteNicki commented on src/lightning/pytorch/CHANGELOG.md:
```diff
@@ -10,6 +10,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
+- WeightAveraging callback that wraps the PyTorch AveragedModel class ([#20545](https://github.com/Lightning-AI/pytorch-lightning/pull/20545))
+
+
+- Added support for general mappings being returned from `training_step` when using manual optimization ([#21011](https://github.com/Lightning-AI/pytorch-lightning/pull/21011))
+
+
```
```suggestion
```

SkafteNicki commented:
@Borda should be good now

kzrpg commented on src/lightning/pytorch/callbacks/weight_averaging.py:
```diff
@@ -0,0 +1,363 @@
+            if is_overridden("configure_model", pl_module):
+                rank_zero_warn(
+                    "You're using the WeightAveraging callback with a model that overrides the configure_model "
+                    "callback. WeightAveraging doesn't support sharding model layers, so you may run out of memory."
```
What changes do we need to support sharded models like when using deepspeed/FSDP?

npuichigo replied:
Any update on this?

catalpaaa commented:
Congrats <3

amorehead commented:
One small caveat for folks to be aware of when using this callback: If your `LightningModule` reaches the end of your test epoch and then calls `on_train_end()`, the callback will swap your current (unaveraged) model weights with your EMA (averaged) model weights prior to (potentially) saving a final model checkpoint. If such a checkpoint is saved, it will contain your averaged model weights, not the unaveraged weights. This is fine for most downstream use cases, assuming the model is truly finished training. However, if you later decide to load such a checkpoint and finetune the model for additional training epochs, your training losses will likely look fine, but your validation/test performance might suffer.

I've seen this myself, where loading a checkpoint after it has reached its `trainer.max_epochs=20000` and then manually training for longer (after setting a larger value of `trainer.max_epochs=40000`) resulted in normal training losses with no disruption, yet I noticed a huge drop in validation performance as shown in the image below. This is likely because, from then on, the callback is treating the averaged weights as the unaveraged model weights.
<img width="1132" height="654" alt="image" src="https://github.com/user-attachments/assets/299bbcb8-0de8-48a5-9ce3-35cee8b4c306" />

philgzl commented:
Thanks. So if I understand correctly this is all because the non-averaged weights are not saved at the end of training.

This seems like a quite drastic behavior for just reducing the checkpoint size. Would it make sense to at least add an option to save the non-averaged weights, so users can resume training or inspect the non-averaged weights? I would personally use that option by default in all my projects.

Just to be clear, both the averaged and non-averaged weights are saved during training such that training can be correctly resumed in case of a crash, right?

What about the interaction with a `ModelCheckpoint` which tracks e.g. validation loss? Is a best checkpoint saved mid-training considered ready for "production", as in it does not contain the non-averaged weights? Or do I need to manually to load the averaged weights after loading such a checkpoint?
